crates/wasmtime/src/runtime/component/func/host.rs

```diff
@@ -289,15 +288,45 @@ fn validate_inbounds<T: ComponentType>(memory: &[u8], ptr: &ValRaw) -> Result<us
     Ok(ptr)
 }
 
-unsafe fn handle_result(func: impl FnOnce() -> Result<()>) {
-    match crate::runtime::vm::catch_unwind_and_longjmp(func) {
+unsafe fn call_host_and_handle_result<T>(
+    cx: *mut VMOpaqueContext,
+    func: impl FnOnce(
+        *mut ComponentInstance,
+        &Arc<ComponentTypes>,
+        StoreContextMut<'_, T>,
+    ) -> Result<()>,
+) {
+    let cx = VMComponentContext::from_opaque(cx);
+    let instance = (*cx).instance();
+    let types = (*instance).component_types();
+    let raw_store = (*instance).store();
+    let store = StoreContextMut::from_raw(raw_store);
+
+    let res = crate::runtime::vm::catch_unwind_and_longjmp(|| {
+        if let Err(e) = store.0.call_hook(CallHook::CallingHost) {
+            return Err(e);
+        }
+
+        let res = func(instance, types, store);
+
+        let store = StoreContextMut::<T>::from_raw(raw_store);
```

Comment:
To avoid needing to call this twice could the above use `store.as_context_mut()` as an argument? That should effectively reborrow to allow `store` to be used here again